[ML] Adding advanced job wizard

jgowdyelastic opened the pull request:
The advanced new job wizard is now no long a separate page, it is a "mode" in the new job wizards. Allowing for the reuse of components added for the other wizards.

Notable differences compared to other wizards:
- No time range selection. The when the job is created, the user is taken back to the jobs management page to start the datafeed. This is the same behaviour as before.
- Editable JSON. On the datafeed and pick fields steps of the wizard, the user is able to edit the underling JSON configurations. This may also be allowed on the summary step.
- No charts or running job progress. In keeping with the behaviour of the original wizard.


![2019-10-08 17-29-06 2019-10-08 17_36_14](https://user-images.githubusercontent.com/22172091/66414911-42961400-e9f2-11e9-84b4-f42d804a96aa.gif)

This is the initial version of this feature and will be followed up with further enhancements.
The project tracking outstanding items is here https://github.com/elastic/kibana/projects/30

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

~~- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)~~
~~- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)~~
~~- [ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials~~
~~- [ ] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios~~
~~- [ ] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)~~

### For maintainers

~~- [ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)~~
~~- [ ] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)~~

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/1275/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/1306/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/1904/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2052/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2303/)
* Commit: 89d5cb04948f01f0d6d10eb3f813b31b9e385d17

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2551/)
* Commit: 4e1e755958ebb186b150c4694903e77b01ecb121

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/2763/)
* Commit: 97500ffdeccf334319a2a5e6ab3d7748b1a892ab

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3026/)
* Commit: 55d2d119a51b5083d5cdf238c778b1396f89f663

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3049/)
* Commit: 1023ed9ce735bc1e4e4ad4cfce6cc57b1daaa126

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3104/)
* Commit: 41e15cb84d573639923c04943964e4711ea7fb75

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3117/)
* Commit: 52609d254f55009f3d1c2cb721c858d666ab21d2

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3193/)
* Commit: 67b91ed9a2e8695b76e5191b866a79ab45e4af10

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3215/)
* Commit: 6572afcf273d21641e248c1a4199521303e91938

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3324/)
* Commit: aa3fe4903a1d123190a1bed2e70342746e00fa77

jgowdyelastic commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3326/)
* Commit: aa3fe4903a1d123190a1bed2e70342746e00fa77

jgowdyelastic commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3329/)
* Commit: aa3fe4903a1d123190a1bed2e70342746e00fa77

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3355/)
* Commit: 8b73cfe0bfba2ab0f2b9499cb52298d64af9419c

pheyos commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/new_job/wizard_steps.tsx:
```diff
@@ -0,0 +1,230 @@
+      )}
+      {currentStep === WIZARD_STEPS.ADVANCED_CONFIGURE_DATAFEED && (
+        <Fragment>
+          <Title data-test-subj="mlJobWizardStepTitlePickFields">
```
This should be somethink like `mlJobWizardStepTitleConfigureDatafeed`

pheyos commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/pick_fields_step/components/summary_count_field/summary_count_field_select.tsx:
```diff
@@ -0,0 +1,51 @@
+      selectedOptions={selection}
+      onChange={onChange}
+      isClearable={true}
+      data-test-subj="summaryCountFieldNameSelect"
```
Would be good to prefix all of our data-test-subj attributes with `ml`

jgowdyelastic commented:
> Usability: in the pick fields step there's no indicator that a detector is needed to make the Next button active

Fixed in 103e732

> Regression: model memory limit is prefilled with 10MB (was 1GB)

Fixed in f662603

> Regression: the detector function selection offers way less options (15) than in the old wizard (35)

Currently working on this

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3456/)
* Commit: 103e7328f6da572095afd36b95ba2d821ced2eeb

peteharverson commented on x-pack/legacy/plugins/ml/common/util/job_utils.js:
```diff
@@ -438,6 +439,31 @@ export function basicJobValidation(job, fields, limits, skipMmlChecks = false) {
+  const valid = true;
+
+  if (datafeed) {
+    // if (_.isEmpty(datafeed.query)) {
```
Has this been left in for a reason?

jgowdyelastic replied:
yes, datafeed validation is an outstanding issue.
https://github.com/elastic/kibana/projects/30#card-27550795

peteharverson commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/common/job_validator/util.ts:
```diff
@@ -151,6 +151,20 @@ export function populateValidationMessages(
+    basicValidations.query.message = msg;
+  } else if (validationResults.contains('query_invalid')) {
+    basicValidations.query.valid = false;
+    const msg = i18n.translate('xpack.ml.newJob.wizard.validateJob.queryIsInvalidJSON', {
```
Nit - would something like `queryIsInvalidElasticsearchQuery` be better than `queryIsInvalidJSON` for this i18n message?

peteharverson commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/pick_fields_step/components/advanced_detector_modal/advanced_detector_modal.tsx:
```diff
@@ -0,0 +1,308 @@
+  const [fieldOptionEnabled, setFieldOptionEnabled] = useState(true);
+
+  const aggOptions: EuiComboBoxOptionProps[] = aggs.map(createAggOption);
+  // .sort((a, b) => a.label.localeCompare(b.label));
```
Is this needed?

peteharverson commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/pick_fields_step/components/advanced_detector_modal/advanced_detector_modal.tsx:
```diff
@@ -0,0 +1,308 @@
+  }
+  return [
+    {
+      label: 'mlcategory',
```
Is the `mlcategory` field name stored in any constants file?

peteharverson commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/pick_fields_step/components/advanced_view/detector_list.tsx:
```diff
@@ -0,0 +1,111 @@
+  return (
+    <Fragment>
+      <EuiTitle size="xs">
+        <h3>Detectors</h3>
```
Needs internationalizing

peteharverson commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/pick_fields_step/components/advanced_view/detector_list.tsx:
```diff
@@ -0,0 +1,111 @@
+                        color="primary"
+                        onClick={() => onEditJob(i)}
+                        iconType="pencil"
+                        aria-label="Edit"
```
Needs in18n

peteharverson commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/pick_fields_step/components/advanced_view/detector_list.tsx:
```diff
@@ -0,0 +1,111 @@
+                        color="danger"
+                        onClick={() => onDeleteJob(i)}
+                        iconType="trash"
+                        aria-label="Delete"
```
Needs in18n

peteharverson commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/pick_fields_step/components/advanced_detector_modal/advanced_detector_modal.tsx:
```diff
@@ -0,0 +1,308 @@
+  }
+
+  return (
+    <ModalWrapper onCreateClick={onCreateClick} closeModal={closeModal} saveEnabled={saveEnabled()}>
```
I can't see any way to edit the description of the detector? Is it hidden away somewhere? It used to be at the top of the Add detector modal.

jgowdyelastic replied:
Will add in a follow up PR.
https://github.com/elastic/kibana/projects/30#card-27607228

walterra commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/datafeed_step/components/frequency/frequency_input.tsx:
```diff
@@ -0,0 +1,56 @@
+  );
+  const [validation, setValidation] = useState(jobValidator.frequency);
+  const [frequency, setFrequency] = useState(
+    jobCreator.frequency === null ? '' : jobCreator.frequency
```
Might be worth creating a function for `jobCreator.frequency === null ? '' : jobCreator.frequency` it's used similarly multiple times in this file.

walterra commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/datafeed_step/components/query/description.tsx:
```diff
@@ -0,0 +1,24 @@
+import { i18n } from '@kbn/i18n';
+import { EuiFormRow } from '@elastic/eui';
+
+interface Props {
```
Can you try if you don't need this custom prop interface at all? I think I remember `FC` already has `children` as a prop and extends what you pass in.

peteharverson commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/pick_fields_step/components/bucket_span_estimator/bucket_span_estimator.tsx:
```diff
@@ -15,14 +16,23 @@ interface Props {
+
   return (
-    <EuiButton disabled={status === ESTIMATE_STATUS.RUNNING} onClick={estimateBucketSpan}>
+    <EuiButton
```
The estimate bucket span is returning an error for me - looks like the time range has defaulted to last 15 mins in the URL.

![image](https://user-images.githubusercontent.com/7405507/66560143-a9304480-eb4e-11e9-8350-c406413f128b.png)

jgowdyelastic replied:
The full time range for the index should probably be used here. There is a card in the project to  investigate this.
https://github.com/elastic/kibana/projects/30#card-27509086

walterra commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/datafeed_step/datafeed.tsx:
```diff
@@ -0,0 +1,64 @@
+export const DatafeedStep: FC<StepProps> = ({ setCurrentStep, isCurrentStep }) => {
+  const { jobValidator, jobValidatorUpdated } = useContext(JobCreatorContext);
+  const [nextActive, setNextActive] = useState(false);
+  const [validQuery, setValidQuery] = useState(false);
```
Nit: Maybe `isValidQuery` makes it more clear it's a `boolean`. Since it's passed on to another component and seeing it there I wasn't sure if it's a flag or if it contains a valid query.

walterra commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/datafeed_step/components/query_delay/description.tsx:
```diff
@@ -0,0 +1,43 @@
+import { Validation } from '../../../../../common/job_validator';
+
+interface Props {
+  children: JSX.Element;
```
Related to the other comment about `Props`, I think you don't need `children` here, `FC` should already consider it by itself.

walterra commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/datafeed_step/components/query_delay/query_delay_input.tsx:
```diff
@@ -0,0 +1,47 @@
+  );
+  const [validation, setValidation] = useState(jobValidator.queryDelay);
+  const [queryDelay, setQueryDelay] = useState(
+    jobCreator.queryDelay === null ? '' : jobCreator.queryDelay
```
Similar to the other comment, maybe `jobCreator.queryDelay === null ? '' : jobCreator.queryDelay` could go into a function and be reused.

walterra commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/datafeed_step/components/scroll_size/description.tsx:
```diff
@@ -0,0 +1,43 @@
+import { Validation } from '../../../../../common/job_validator';
+
+interface Props {
+  children: JSX.Element;
```
again, `children` might not be necessary here.

walterra commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/job_details_step/components/advanced_section/advanced_section.tsx:
```diff
@@ -24,28 +32,49 @@ interface Props {
+        </EuiFlexGroup>
+      </Fragment>
+    );
+  } else {
```
nit: you could skip the `else` here and just `return`.

walterra commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/pick_fields_step/components/categorization_field/categorization_field.tsx:
```diff
@@ -0,0 +1,56 @@
+
+export const CategorizationField: FC = () => {
+  const { jobCreator: jc, jobCreatorUpdate, jobCreatorUpdated } = useContext(JobCreatorContext);
+  if (
```
AFAIK we must not return before all hooks are called in a component, I wonder if you get console warnings once this component updates? Maybe it can be solved without refactored this whole component by just wrapping it in a `WithJobCreator` component. You would do the check and return `null` in the wrapper and only render this component with `jobCreator` being always available and calling all hooks.

walterra commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/pick_fields_step/components/influencers/influencers.tsx:
```diff
@@ -14,16 +14,22 @@ import {
 export const Influencers: FC = () => {
   const { jobCreator: jc, jobCreatorUpdate, jobCreatorUpdated } = useContext(JobCreatorContext);
-  if (isMultiMetricJobCreator(jc) === false && isPopulationJobCreator(jc) === false) {
+  if (
```
As mentioned in the other component, returning `null` here before calling all hooks isn't valid. All hooks need to be called on every render.

walterra commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/pick_fields_step/components/summary_count_field/summary_count_field.tsx:
```diff
@@ -0,0 +1,56 @@
+
+export const SummaryCountField: FC = () => {
+  const { jobCreator: jc, jobCreatorUpdate, jobCreatorUpdated } = useContext(JobCreatorContext);
+  if (
```
Again, as mentioned in the other component, returning `null` here before calling all hooks isn't valid. All hooks need to be called on every render.

walterra commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/index_or_search/preconfigured_job_redirect.ts:
```diff
@@ -0,0 +1,47 @@
+  let page = '';
+
+  if (created === CREATED_BY_LABEL.SINGLE_METRIC) {
+    page = 'single_metric';
```
nit: maybe worth moving these strings to constants or an enum?

peteharverson commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/new_job/wizard_steps.tsx:
```diff
@@ -0,0 +1,230 @@
+  jobType: JOB_TYPE;
+}
+
+export const WizardSteps: FC<Props> = ({
```
The numbers at the top look like links, but clicking on them only works if you've moved forwards to that step, and back. Maybe the links should only become links if it's possible to jump to that step?

peteharverson commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/pick_fields_step/components/advanced_detector_modal/advanced_detector_modal.tsx:
```diff
@@ -0,0 +1,308 @@
+
+const excludeFrequentOptions: EuiComboBoxOptionProps[] = [{ label: 'all' }, { label: 'none' }];
+
+export const AdvancedDetectorModal: FC<Props> = ({
```
Should any script_fields set in the datafeed step be available in the detector modal? I configured `script_fields` but the field I added didn't seem to be available when configuring a detector.

jgowdyelastic replied:
yep!
https://github.com/elastic/kibana/projects/30#card-27612562

darnautov commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/new_job/wizard_steps.tsx:
```diff
@@ -0,0 +1,230 @@
+    return '';
+  }
+
+  const stepsConfig = [
```
can be moved out of component (need to replace `onClick` with `nextStep`)

darnautov commented on x-pack/legacy/plugins/ml/public/ml_nodes_check/check_ml_nodes.ts:
```diff
@@ -0,0 +1,47 @@
+    if (nodes.count !== undefined && nodes.count > 0) {
+      Promise.resolve();
+    } else {
+      throw Error;
```
an error message might be useful

darnautov commented on x-pack/legacy/plugins/ml/public/ml_nodes_check/check_ml_nodes.ts:
```diff
@@ -0,0 +1,47 @@
+export async function getMlNodeCount() {
+  try {
+    const nodes = await ml.mlNodeCount();
+    mlNodeCount = nodes.count;
```
side effect in a getter function 👀

jgowdyelastic commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/pick_fields_step/components/influencers/influencers.tsx:
```diff
@@ -14,16 +14,22 @@ import {
 } from '../../../../../common/job_creator';
 import { Description } from './description';
 
 export const Influencers: FC = () => {
```
good point.
this can be added, but it might be worth reviewing the need for this.
What would the user need to add that isn't in the field list?

peteharverson replied:
You can't add `mlcategory`, which is useful when creating categorization jobs. This needs to be available.

jgowdyelastic replied:
added...

jgowdyelastic commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/pick_fields_step/components/summary_count_field/summary_count_field.tsx:
```diff
@@ -0,0 +1,56 @@
+} from '../../../../../common/job_creator';
+import { Description } from './description';
+
+export const SummaryCountField: FC = () => {
```
this was copied from the tooltip.
@sophiec20 do you have a suggestion for better text here?

peteharverson replied:
How about something like `Optional, for use if input data has been pre-summarized. The name of the field that contains the count of raw data points.` ?

jgowdyelastic commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/datafeed_step/datafeed.tsx:
```diff
@@ -0,0 +1,64 @@
+import { JobCreatorContext } from '../job_creator_context';
+import { JsonEditorFlyout, EDITOR_MODE } from '../common/json_editor_flyout';
+
+export const DatafeedStep: FC<StepProps> = ({ setCurrentStep, isCurrentStep }) => {
```
yes!
https://github.com/elastic/kibana/projects/30#card-27610524

jgowdyelastic commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/job_details_step/job_details.tsx:
```diff
@@ -55,13 +55,12 @@ export const JobDetailsStep: FC<Props> = ({
           />
-          <EuiSpacer />
+
           <AdvancedSection
```
yes this can be added in a follow up.
https://github.com/elastic/kibana/projects/30#card-27610582

jgowdyelastic commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/common/job_validator/job_validator.ts:
```diff
@@ -31,6 +31,10 @@ export interface BasicValidations {
+  scrollSize: Validation;
 }
 
 export class JobValidator {
```
Yes this has been missed.
https://github.com/elastic/kibana/projects/30#card-27610610

jgowdyelastic commented on x-pack/legacy/plugins/ml/common/util/job_utils.js:
```diff
@@ -438,6 +439,31 @@ export function basicJobValidation(job, fields, limits, skipMmlChecks = false) {
+    //   valid = false;
+    // } else {
+    //   messages.push({ id: 'query_valid' });
+    // }
```
datafeed validation is coming in a future PR, i'd like to leave this code commented out until then.
https://github.com/elastic/kibana/projects/30#card-27550795

jgowdyelastic commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/common/job_creator/job_creator.ts:
```diff
@@ -245,6 +245,70 @@ export class JobCreator {
+    if (fieldName !== null) {
+      this._job_config.analysis_config.summary_count_field_name = fieldName;
+    } else {
+      delete this._job_config.analysis_config.summary_count_field_name;
```
these setters might change in the future if we're now not supposed to use them.
These setters are not just basic setters,  they create properties deep inside the job or datafeed configs. sometimes they'll do some house keeping by removing the outer object if it is now empty.
Changing them to be standard functions might make this clearer.

jgowdyelastic commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/common/job_creator/job_creator.ts:
```diff
@@ -245,6 +245,70 @@ export class JobCreator {
+    }
+
+    const filters = this._job_config.analysis_config.categorization_filters;
+    if (filters.includes(filter) === false) {
```
i guess it's a matter of taste. i prefer writing boolean checks this way rather than say 
`if (!filters.includes(filter)) {`
as i think it is more obvious. `!` chars can be missed.

jgowdyelastic commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/datafeed_step/components/frequency/description.tsx:
```diff
@@ -0,0 +1,43 @@
+      description={
+        <FormattedMessage
+          id="xpack.ml.newJob.wizard.datafeedStep.frequency.description"
+          defaultMessage=" The interval between searches."
```
good eyes!

jgowdyelastic commented on x-pack/legacy/plugins/ml/public/jobs/new_job_new/pages/components/datafeed_step/components/query/query_input.tsx:
```diff
@@ -0,0 +1,62 @@
+
+  useEffect(() => {
+    const validJson = isValidJson(queryString);
+    setValidQuery(validJson);
```
sorry, i don't understand this comment?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3711/)
* Commit: db499e34c5e206818cb189a822c4542cc1b7639e

jgowdyelastic commented:
retest

jgowdyelastic commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3798/)
* Commit: b458532cb46dfbe475029fd76ff9fec659e9bba3

jgowdyelastic commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3880/)
* Commit: b458532cb46dfbe475029fd76ff9fec659e9bba3

jgowdyelastic commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3901/)
* Commit: dde7fb27f5cee64828760d9fb6457c866e1b6d22

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3919/)
* Commit: 5755719753665dd82b3549eef804f0c88f5d4dd9

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/3947/)
* Commit: fd61fbf7cadd26b09d309ae2d958c27696faa9ae

jgowdyelastic commented:
https://github.com/elastic/kibana/issues/18374
